### PR TITLE
changes in the way of initialization Project from dict and pandas df

### PR DESCRIPTION
**Pull request by Khoroshevskyi**
Address #402
Also addresses #401 

Big changes in functions:
- to_dict (returning only crucial object variables. This method is more efficient, than returning all variables)
- from_dict (initialization of the object from dict of variables that were generated by to_dict function)
- from_pandas (added possibility to initialize project also by using subsample table and config file )
- new equal function (Now equal function will check only samples. - previous version was checking all variables. If any path to the file was not the same it would return that this projects were not the same. Now it's fixed)

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -15,6 +15,10 @@ Way of initialization project from dictionary. Now it's possible as follows: Pro
 
 - Feature of initializing `peppy.Project` from `pandas.DataFrame`
 
+## [0.34.0] -- 2022-08-17
```
Comment (rafalstepien) [inline docs/changelog.md:18]: Shouldn't be `0.35.0`? `0.34.0` is the current one

File: peppy/project.py [hunk 2]
```diff
@@ -160,13 +160,10 @@ def __init__(
...
-        dict_self = self._convert_to_dict(self)
-        dict_other = self._convert_to_dict(other)
+        samples_self = [s.to_dict() for s in self.samples]
+        samples_other = [s.to_dict() for s in other.samples]
 
-        dict_self["_samples"] = [s.to_dict() for s in self.samples]
-        dict_other["_samples"] = [s.to_dict() for s in other.samples]
-
-        return dict_self == dict_other
+        return samples_self == samples_other
```
Comment (rafalstepien) [inline peppy/project.py:166]: You can go direclty for
```python
def __eq__(self, other):
    return [s.to_dict() for s in self.samples] == [s.to_dict() for s in other.samples]
```

File: tests/smoketests/test_Project.py [hunk 3]
```diff
@@ -528,6 +527,7 @@ def test_subtable(self, example_pep_cfg_path):
         Verify that the sample merging takes place
         """
         p = Project(cfg=example_pep_cfg_path)
+        a = p.samples
```
Comment (rafalstepien) [inline tests/smoketests/test_Project.py:530]: Maybe something more verbose than `a` as variable name? :D

Comment (Khoroshevskyi) [reply tests/smoketests/test_Project.py:530]: we don't need it. I forgot to delete this line

File: docs/changelog.md (hunk 1 above)
Comment (Khoroshevskyi) [reply docs/changelog.md:18]: I think, this mistake is after merging branches, I'll fix it

**Review by rafalstepien — APPROVED** [review]

File: peppy/project.py [hunk 4]
```diff
@@ -356,19 +347,27 @@ def _get_table_from_samples(self, index, initial=False):
         df.set_index(keys=index, drop=False, inplace=True)
         return df
 
-    def parse_config_file(self, cfg_path, amendments=None):
+    def parse_config_file(
+        self,
+        cfg_path: str = None,
+        amendments: Iterable[str] = None,
+        config: dict = None,
```
Comment (nsheff) [inline peppy/project.py:354]: This should be a separate function, I think.

The function signature like this is not clear. This is similar to what we've discussed before -- this seems like there should be two functions, with different signatures, one that calls the other

**Review by nsheff — CHANGES_REQUESTED** [review]

File: peppy/project.py (hunk 4 above)
Comment (Khoroshevskyi) [reply peppy/project.py:354]: Thanks for mentioning it! I think, I am even not using this argument! I have to check it again manually

**Comment by nsheff** [comment]
Is the purpose of this to simplify inserting into the database with pepdbagent?

I mean, what prompted the change?

**Comment by Khoroshevskyi** [comment]
> Is the purpose of this to simplify inserting into the database with pepdbagent?

Not simplifying, I would say make it more efficient (especially storage efficient) and more reasonable to store one sample table, then 3 sample tables in different configurations. I assume, now it will take 3x less space in pepdb.

**Comment by nsheff** [comment]
> 

and will this also make it possible to use simplified projects, because they are unprocessed? like, I mean collapsed constant rows to the project level?

**Comment by Khoroshevskyi** [comment]
> > 
> 
> and will this also make it possible to use simplified projects, because they are unprocessed? like, I mean collapsed constant rows to the project level?

Do you mean projects, such as geofetch is produsing (e.g. where config file contains append)?  - then Yes!

File: peppy/project.py (hunk 4 above)
Comment (nsheff) [reply peppy/project.py]: What are you going to do about this?

**Comment by nsheff** [comment]
> Do you mean projects, such as geofetch is producing (e.g. where config file contains append)? - then Yes!

Yes, that's what I mean. Great!


Can you rename this Pull Request to something more clear? Why does it say "402" ?

**Comment by rafalstepien** [comment]
> > Do you mean projects, such as geofetch is producing (e.g. where config file contains append)? - then Yes!
> 
> Yes, that's what I mean. Great!
> 
> Can you rename this Pull Request to something more clear? Why does it say "402" ?

I think because it originated from issue number 402

**Comment by nsheff** [comment]
> I think because it originated from issue number 402

Indeed. @Khoroshevskyi you have to use the # sign to link to the issue and indicate it as an issue number. Also, this probably belongs in the description of the PR

**Comment by rafalstepien** [comment]
So after changing the name of PR are we good to go?

**Comment by nsheff** [comment]
I made 1 request for a change, this still is not addressed (see above)

**Comment by Khoroshevskyi** [comment]
> > Do you mean projects, such as geofetch is producing (e.g. where config file contains append)? - then Yes!
> 
> Yes, that's what I mean. Great!
> 
> Can you rename this Pull Request to something more clear? Why does it say "402" ?

As Rafal mentioned above, the name was originated from the issue number 402. All 4 changes are related to this problem. `eq` function is related to functionality `to_dict`; `from_dict` is dependent from `to_dict` ; And `from_pandas` has similar function as `from_dict` . That's why all this changes were made here. Is there something else to change @nsheff ?

**Review by nsheff — APPROVED** [review]

**Comment by Khoroshevskyi** [comment]
I have chenged the names of the dict keys, could you check it please? @nsheff  --> https://github.com/pepkit/peppy/pull/405/commits/802ea1a112635f1f233a5faa12e0716bf13d8157

**Pull request merged**